src/components/MachineSelector/Configurator/MachineConfigurator.tsx

```diff
@@ -0,0 +1,337 @@
+import CloseIcon from '@material-ui/icons/Close';
+import Typography from '@material-ui/core/Typography';
+import TextField from '@material-ui/core/TextField';
+import Grid from '@material-ui/core/Grid';
+import Tooltip from '@material-ui/core/Tooltip';
+import CircularProgress from '@material-ui/core/CircularProgress';
+import BuildIcon from '@material-ui/icons/Build';
+import AddCircleOutlineIcon from '@material-ui/icons/AddCircleOutline';
+import HighlightOffIcon from '@material-ui/icons/HighlightOff';
+
+import './MachineConfigurator.css';
+import { Dispatch, bindActionCreators } from 'redux';
+import { RootAction } from 'StoreTypes';
+import { connect } from 'react-redux';
+import { setAlerts, updateAssets } from 'store/actions/root-action';
+import { getMachineIds, saveMachineIds } from 'services/appCRUD/appConfService';
+import { validateAsset } from './assetValidator';
+
+const styles = (theme: Theme) =>
+  createStyles({
+    root: {
+      margin: 0,
+      padding: theme.spacing(2),
+    },
+    closeButton: {
+      position: 'absolute',
+      right: theme.spacing(1),
+      top: theme.spacing(1),
+      color: theme.palette.grey[500],
+    },
+  });
+
+export interface DialogTitleProps extends WithStyles<typeof styles> {
+  id: string;
+  children: React.ReactNode;
+  onClose: () => void;
+}
+
+/**
+ * This component is used as machine configurator and enabled only in settings page for admin users
+ */
+
+/**
+ * Title component for machine configuration popup
+ */
+const DialogTitle = withStyles(styles)((props: DialogTitleProps) => {
+  const { children, classes, onClose, ...other } = props;
+  return (
+    <MuiDialogTitle disableTypography className={classes.root} {...other}>
+      <Typography variant="h6">{children}</Typography>
+      {onClose ? (
+        <IconButton
+          aria-label="close"
+          className={classes.closeButton}
+          onClick={onClose}
+        >
+          <CloseIcon />
+        </IconButton>
+      ) : null}
+    </MuiDialogTitle>
+  );
+});
+
+type FormField = {
+  assetExternalId: string;
+};
+
+const defaultValues = {
+  assetExternalId: '',
+};
+
+/**
+ * This is the main functional component having the machine configurator
+ */
+const MachineConfigurator: FC<Props> = (props: Props) => {
+  const [open, setOpen] = useState(false);
+  const [assetExternalIds, setAssetExternalIds] = useState<string[]>([]);
+  const [assetValidating, setAssetValidating] = useState(false);
+
+  const onError = (msg: string) => {
+    props.setAlerts({
+      type: 'error',
+      text: msg,
+      hideApp: false,
+    });
+  };
+
+  const { handleSubmit, control, reset } = useForm({
+    defaultValues,
+  });
+
+  /**
+   * This will open the configurator popup on machine settings icon click.
+   * restoreMachineConfig function is used to fill the form fields with previously saved values
+   */
+  const handleClickOpen = () => {
+    setOpen(true);
+    setTimeout(() => {
```

Comment:
Whats the purpose of delaying this?

Reply:
This was actually used in a previous implementation and its no longer required to have a timeout now. Removed it